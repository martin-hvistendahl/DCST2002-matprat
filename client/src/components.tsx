import * as React from 'react';
import { Component } from 'react-simplified';
import { Alert, Card, Row, Column, Form, Button, RecipeView } from './widgets';
import { NavLink, Redirect } from 'react-router-dom';
import service, { Country, Category, Ingredient, Recipe, Recipe_Content } from './service';
import { createHashHistory } from 'history';

const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a student

/**
 * Renders form to create new recipe.
 */
export class NewRecipe extends Component {
  countries: Country[] = [];
  categories: Category[] = [];
  ingredients: Ingredient[] = [];
  recipe_content: Recipe_Content[] = [];
  ingredient: string = '';

  name: string = '';
  description: string = '';
  steps: string = '';
  portions: number = 0;
  picture_adr: string = '';
  category_id: number = 0;
  category_name: string = '';
  country_id: number = 0;
  country_name: string = '';

  render() {
    console.log(this.categories, this.countries);
    return (
      <>
        <Card title="Registrer en ny oppskrift">
          {/* input navn */}
          <Column>
            <Column width={2}>
              <Form.Label>Name:</Form.Label>
            </Column>
            <Column>
              <Form.Input
                type="text"
                value={this.name}
                onChange={(event) => (this.name = event.currentTarget.value)}
              />
            </Column>
          </Column>
          {/* input beksrivelse */}
          <Column>
            <Column width={2}>
              <Form.Label>Description:</Form.Label>
            </Column>
            <Column>
              <Form.Textarea
                style={{ width: '300px' }}
                type="text"
                value={this.description}
                onChange={(event) => (this.description = event.currentTarget.value)}
                rows={5}
              />
            </Column>
          </Column>
          {/* input steg */}
          <Column>
            <Column width={2}>
              <Form.Label>Steg:</Form.Label>
            </Column>
            <Column>
              <Form.Textarea
                style={{ width: '600px' }}
                type="text"
                value={this.steps}
                onChange={(event) => (this.steps = event.currentTarget.value)}
                rows={10}
              />
            </Column>
          </Column>
          {/* input antall porsjoner */}
          <Column>
            <Column width={2}>
              <Form.Label>Porjsoner:</Form.Label>
            </Column>
            <Column>
              <Form.Input
                type="number"
                value={this.portions}
                onChange={(event) => (this.portions = event.currentTarget.value)}
              />
            </Column>
          </Column>
          {/* input bilde url */}
          <Column>
            <Column width={2}>
              <Form.Label>Bilde url:</Form.Label>
            </Column>
            <Column>
              <Form.Input
                type="text"
                value={this.picture_adr}
                onChange={(event) => (this.picture_adr = event.currentTarget.value)}
              />
            </Column>
          </Column>
          {/* velg land retten kommer fra */}
          <Column>
            <Column width={2}>
              <Form.Label>Land:</Form.Label>
            </Column>
            <Column>
              <select
                id="choseCountry"
                onChange={() => {
                  this.checkCountry(event?.target.value);
                }}
              >
                {/* feilmeldingen kommer fra selected under */}
                {this.countries.map((country: Country, i: number) => (
                  <option key={country.land_id} value={country.land_id}>
                    {country.land_navn}
                  </option>
                ))}
                <option value="0">ikke på liste</option>
              </select>
              <Form.Input
                id="addCountry"
                style={{ opacity: '0', width: '300px' }}
                type="text"
                value={this.country_name}
                onChange={(event) => (this.country_name = event.currentTarget.value)}
                placeholder="Skriv inn landet retten kommer fra"
              ></Form.Input>
              <Button.Success
                id="addCountryBtn"
                onClick={() => {
                  this.addCountryFunc();
                }}
              >
                Legg til
              </Button.Success>
              {/* må lage select og options som cars */}
            </Column>
          </Column>
          {/* velg retten sin kategori */}
          <Column>
            <Column width={2}>
              <Form.Label>Kategori:</Form.Label>
            </Column>
            <Column>
              <select
                id="choseCategory"
                onChange={() => {
                  this.checkCategory(event?.target.value);
                }}
              >
                {this.categories.map((category: Category) => (
                  <option key={category.kategori_id} value={category.kategori_id}>
                    {category.kategori_navn}
                  </option>
                ))}
                <option value="0">ikke på liste</option>
              </select>
              <Form.Input
                id="addCategory"
                style={{ opacity: '0', width: '300px' }}
                type="text"
                value={this.category_name}
                onChange={(event) => (this.category_name = event.currentTarget.value)}
                placeholder="Skriv inn kattegorien retten tilhører"
              ></Form.Input>
              <Button.Success
                id="addCategoryBtn"
                onClick={() => {
                  this.addIngredientFunc();
                }}
              >
                Legg til
              </Button.Success>
              {/* må lage select og options som cars */}
            </Column>
          </Column>
          {/* print ut alle ingrediense som allerede er i databasen */}
          {/* vidre ideer her er at vi setter en viss lengde og bredde på diven og så hvis den overflower så må man bare skulle 
          nedover, her kan vi også implementere et søkefelt etterhvert for ingredienser. */}
          <Column>
            Ingrediensene som allered er lagret,
            <br /> hvis ingrediensen din ikke er her kan du legge den til!
            <br />
            <Column>
              {this.ingredients.map((ingredient) => (
                <Button.Light
                  id={ingredient.ingred_id}
                  key={ingredient.ingred_id}
                  onClick={() => {
                    this.chooseIngredientFunc(ingredient.ingred_id, ingredient.ingred_navn);
                  }}
                >
                  {ingredient.ingred_navn}
                </Button.Light>
              ))}
            </Column>
          </Column>
          {/* legg til ingredienser */}
          <Column>
            <Form.Input
              id="addIngredient"
              type="text"
              style={{ width: '400px' }}
              value={this.ingredient}
              onChange={(event) => (this.ingredient = event.currentTarget.value)}
              placeholder="Legg til ingredienser"
            ></Form.Input>
            <Button.Success
              onClick={() => {
                this.addIngredientFunc();
              }}
            >
              Legg til
            </Button.Success>
          </Column>
          {/* velg hvor mye av hver inngrediense */}
          <Column>
            <h5>
              Klikk på ingrediensene over for at de skal komme hit og du kan velge hvor mye du skal
              ha av hver ingrediens
            </h5>
            <div id="ingreditentList"></div>
          </Column>
        </Card>
        <Button.Success
          onClick={() => {
            this.addRecipe();
          }}
        >
          Send oppskriften opp til the database
        </Button.Success>
      </>
    );
  }
  addRecipe() {
    let recipe: Recipe = {
      oppskrift_id: 0,
      oppskrift_navn: this.name,
      oppskrift_beskrivelse: this.description,
      oppskrift_steg: this.steps,
      ant_pors: this.portions,
      bilde_adr: this.picture_adr,
      kategori_id: this.category_id,
      land_id: this.country_id,
      ant_like: 0,
    };
    service
      .createRecipe(recipe)
      .then((id) => this.addRecipeIngredient(id))
      .catch((error) => Alert.danger('Creating new recipe: ' + error.message));
  }
  addRecipeIngredient(id: number) {
    this.recipe_content.forEach((element) => {
      element.oppskrift_id = id;
    });
    console.log('første console log', this.recipe_content);
    service
      .createRecipeIngredient(this.recipe_content)
      .then(() => history.push('/recipe/' + id))
      .catch((error) => Alert.danger('Creating new recipe: ' + error.message));

    console.log(this.recipe_content);
  }
  chooseIngredientFunc(id: any, name: string) {
    const btn = document.getElementById(id) as HTMLButtonElement | null;
    if (btn != null) {
      btn.disabled = true;
    }
    /* id må være any for å kunne bruke variablen i setattribute. Den forventer to strings så 
    hvis vi hadde deklarert den som number hadde setAttribute blitt sint*/
    // lager en const som legges til i objectet recipe_content
    const add = { oppskrift_id: 0, ingred_id: id, mengde: 0, maleenhet: '' };

    this.recipe_content.push(add);

    // finner index til dette elemetet i objektet, skal brukes senere til å mappe
    // hver enkelt inputfelt til et obejct sin menge eller måleenhet
    const index = this.recipe_content
      .map(function (element) {
        return element.ingred_id;
      })
      .indexOf(id);

    const emFood = document.createElement('em');
    const inputNumberOf = document.createElement('input');
    const inputMeasurment = document.createElement('input');
    const deleteBtn = document.createElement('button');

    emFood.innerHTML = ' <br />' + name;
    emFood.setAttribute('id', 'emFood' + id);

    inputNumberOf.type = 'number';
    inputNumberOf.setAttribute('id', 'inputNumberOf' + id);
    //@ts-ignore
    inputNumberOf.value = this.recipe_content[index].mengde;
    inputNumberOf.onchange = (event) =>
      (this.recipe_content[index].mengde = event.currentTarget.value);

    inputMeasurment.type = 'text';
    inputMeasurment.setAttribute('id', 'inputMeasurment' + id);
    inputMeasurment.value = this.recipe_content[index].maleenhet;
    inputMeasurment.onchange = (event) =>
      (this.recipe_content[index].maleenhet = event.currentTarget.value);

    deleteBtn.innerHTML = 'x';
    deleteBtn.onclick = () => {
      this.recipe_content.splice(index, 1);
      btn != null ? (btn.disabled = false) : '';
      emFood.remove();
      inputNumberOf.remove();
      inputMeasurment.remove();
      deleteBtn.remove();
    };

    document.getElementById('ingreditentList').appendChild(emFood);
    document.getElementById('ingreditentList').appendChild(inputNumberOf);
    document.getElementById('ingreditentList').appendChild(inputMeasurment);
    document.getElementById('ingreditentList').appendChild(deleteBtn);
  }
  addCountryFunc() {
    // sjekker om landet allerede finnes i arrayen med land, hvis ikke legger den til landet i databasen
    // hentet fra databasen
    let isFound = this.countries.some((country) => {
      if (country.land_navn.toLowerCase() == this.country_name.toLowerCase()) {
        return true;
      }
      return false;
    });

    console.log(isFound);
    const result =
      this.country_name.charAt(0).toUpperCase() + this.country_name.slice(1).toLowerCase();
    this.country_name = result;
    if (!isFound && this.country_name != '') {
      service.createCountry(this.country_name).then(() =>
        service
          .getAllCountry()
          .then((countries) => (this.countries = countries))
          .catch((error) => Alert.danger('Error : ' + error.message))
      );
      //sender her bare 1 for at TS skal bli fornøyd, funksjonen nedenfor forventer også et tall
      this.checkCountry(1);
    }
  }
  addIngredientFunc() {
    // sjekker om ingrediensen allerede finnes i arrayen med ingredienser
    // hentet fra databasen
    let isFound = this.ingredients.some((ingredient) => {
      console.log(ingredient.ingred_navn == this.ingredient);
      if (ingredient.ingred_navn == this.ingredient) {
        return true;
      }
      return false;
    });

    console.log(isFound);
    if (!isFound && this.ingredient != '') {
      service.createIngredient(this.ingredient).then(() =>
        service
          .getAllIngredient()
          .then((ingredients) => (this.ingredients = ingredients))
          .catch((error) => Alert.danger('Error : ' + error.message))
      );
    }
  }
  checkCountry(value: number) {
    this.country_id = value;
    const addCountry: any = document.getElementById('addCountry');
    value == 0 ? (addCountry.style.opacity = ' 100') : (addCountry.style.opacity = ' 0');
  }

  checkCategory(value: number) {
    this.category_id = value;
    const addCategory: any = document.getElementById('addCategory');
    value == 0 ? (addCategory.style.opacity = ' 100') : (addCategory.style.opacity = ' 0');
  }
  mounted() {
    service
      .getAllCountry()
      .then((countries) => (this.countries = countries))
      .catch((error) => Alert.danger('Error getting countries: ' + error.message));
    service
      .getAllCategory()
      .then((categories) => (this.categories = categories))
      .catch((error) => Alert.danger('Error getting categories: ' + error.message));
    service
      .getAllIngredient()
      .then((ingredients) => (this.ingredients = ingredients))
      .catch((error) => Alert.danger('Error getting categories: ' + error.message));
  }
}

export class ShowRecipe extends Component<{ match: { params: { id: number } } }> {
  recipe: Recipe = {
    oppskrift_id: 0,
    oppskrift_navn: '',
    oppskrift_beskrivelse: '',
    oppskrift_steg: '',
    ant_pors: 0,
    bilde_adr: '',
    kategori_id: 0,
    land_id: 0,
    ant_like: 0,
  };
  portions: number = 0;
  recipeContent: Recipe_Content[] = [];
  ingredients: Ingredient[] = [];

  render() {
    {
      console.log(this.recipe);
      console.log(this.recipeContent);
      console.log(this.ingredients);
    }

    return (
      <div>
        <Card>
          <img src={this.recipe.bilde_adr}></img>
          <h1>{this.recipe.oppskrift_navn}</h1>
          <p>{this.recipe.oppskrift_beskrivelse}</p>
          <pre>{this.recipe.oppskrift_steg}</pre>
          <h3>Ingredienser</h3>
          Porsjoner <Button.Danger onClick={this.decrementPortions}>-</Button.Danger>{' '}
          <b>{this.portions}</b> <Button.Success onClick={this.incrementPortions}>+</Button.Success>
          {this.recipeContent.map((rc, i) => (
            <p key={i}>
              {i + 1}.{' '}
              {this.ingredients.filter((ing) => rc.ingred_id == ing.ingred_id)[0].ingred_navn}{' '}
              {(rc.mengde * this.portions) / this.recipe.ant_pors} {rc.maleenhet}
            </p>
          ))}
        </Card>
        <Button.Success onClick={() => history.push('/recipe/edit/' + this.recipe.oppskrift_id)}>
          Endre oppskrift
        </Button.Success>
        <Button.Danger onClick={() => this.deleteRecipe(this.recipe.oppskrift_id)}>
          Slett oppskrift
        </Button.Danger>
      </div>
    );
  }

  mounted() {
    service
      .getAllIngredient()
      .then((ingredients) => (this.ingredients = ingredients))
      .catch((error) => Alert.danger('Error getting ingredients: ' + error.message));
    console.log(this.props.match.params.id);
    service
      .getRecipeContent(this.props.match.params.id)
      .then((recipeContent) => (this.recipeContent = recipeContent))
      .catch((error) => Alert.danger('Error getting recipe content: ' + error.message));

    service
      .getRecipe(this.props.match.params.id)
      .then((recipe) => {
        this.recipe = recipe[0];
        this.portions = recipe[0].ant_pors;
      })
      .catch((error) => Alert.danger('Error getting recipe: ' + error.message));
  }
  incrementPortions() {
    this.portions++;
  }
  decrementPortions() {
    if (this.portions > 1) {
      this.portions--;
    }
  }
  deleteRecipe(id: number) {
    service
      .deleteRecipe(id)
      .then(() => history.push('/'))
      .catch((error) => Alert.danger('Error deleting recipe: ' + error.message));
  }
}
export class EditRecipe extends Component<{ match: { params: { id: number } } }> {
  recipe: Recipe = {
    oppskrift_id: 0,
    oppskrift_navn: '',
    oppskrift_beskrivelse: '',
    oppskrift_steg: '',
    ant_pors: 0,
    bilde_adr: '',
    kategori_id: 0,
    land_id: 0,
    ant_like: 0,
  };
  recipeContent: Recipe_Content[] = [];
  ingredients: Ingredient[] = [];

  render() {
    return (
      <>
        <Card title="Endre oppskriften">
          {/* input navn */}
          <Column>
            <Column width={2}>
              <Form.Label>Name:</Form.Label>
            </Column>
            <Column>
              <Form.Input
                type="text"
                value={this.recipe.oppskrift_navn}
                onChange={(event) => (this.recipe.oppskrift_navn = event.currentTarget.value)}
              />
            </Column>
          </Column>
          {/* input beksrivelse */}
          <Column>
            <Column width={2}>
              <Form.Label>Description:</Form.Label>
            </Column>
            <Column>
              <Form.Textarea
                style={{ width: '300px' }}
                type="text"
                value={this.recipe.oppskrift_beskrivelse}
                onChange={(event) =>
                  (this.recipe.oppskrift_beskrivelse = event.currentTarget.value)
                }
                rows={5}
              />
            </Column>
          </Column>
          {/* input steg */}
          <Column>
            <Column width={2}>
              <Form.Label>Steg:</Form.Label>
            </Column>
            <Column>
              <Form.Textarea
                style={{ width: '600px' }}
                type="text"
                value={this.recipe.oppskrift_steg}
                onChange={(event) => (this.recipe.oppskrift_steg = event.currentTarget.value)}
                rows={10}
              />
            </Column>
          </Column>
          {/* input antall porsjoner */}
          <Column>
            <Column width={2}>
              <Form.Label>Porjsoner:</Form.Label>
            </Column>
            <Column>
              <Form.Input
                type="number"
                value={this.recipe.ant_pors}
                //@ts-ignore
                onChange={(event) => (this.recipe.ant_pors = event.currentTarget.value)}
              />
            </Column>
          </Column>
          {/* input bilde url */}
          <Column>
            <Column width={2}>
              <Form.Label>Bilde url:</Form.Label>
            </Column>
            <Column>
              <Form.Input
                type="text"
                value={this.recipe.bilde_adr}
                onChange={(event) => (this.recipe.bilde_adr = event.currentTarget.value)}
              />
            </Column>
          </Column>

          {/* renderer alle ingrediensene som er linket til oppskriften, her kan man også endre på hvor mye det er av hver ingrediens og måleenheten */}
          <Column>
            <div id="outprintIngredient">
              {this.recipeContent.map((rc, i) => (
                <p key={i}>
                  {this.ingredients.filter((ing) => rc.ingred_id == ing.ingred_id)[0].ingred_navn}{' '}
                  <input
                    style={{ width: '50px' }}
                    type="number"
                    value={rc.mengde}
                    onChange={(event) => (
                      //@ts-ignore
                      (rc.mengde = event.currentTarget.value), console.log(this.recipeContent)
                    )}
                  />
                  <input
                    style={{ width: '100px' }}
                    type="text"
                    value={rc.maleenhet}
                    onChange={(event) => (
                      //@ts-ignore
                      (rc.maleenhet = event.currentTarget.value), console.log(this.recipeContent)
                    )}
                  />
                  <Button.Danger
                    onClick={() => this.deleteIngredient(rc.oppskrift_id, rc.ingred_id)}
                  >
                    x
                  </Button.Danger>
                </p>
              ))}{' '}
            </div>
          </Column>
          {/* print ut alle ingrediense som allerede er i databasen */}
          {/* vidre ideer her er at vi setter en viss lengde og bredde på diven og så hvis den overflower så må man bare skulle 
          nedover, her kan vi også implementere et søkefelt etterhvert for ingredienser. */}
          <Column>
            Ingrediensene som allered er lagret,
            <br /> hvis ingrediensen din ikke er her kan du legge den til!
            <br />
            <Column>
              {this.ingredients.map((ingredient) => (
                <>
                  <Button.Light
                    id={ingredient.ingred_id}
                    key={ingredient.ingred_id}
                    onClick={() => {
                      this.addIngredientFunc(ingredient.ingred_id, this.props.match.params.id);
                    }}
                  >
                    {ingredient.ingred_navn}
                  </Button.Light>
                </>
              ))}
            </Column>
          </Column>
        </Card>
        <Button.Success onClick={() => this.pushNewChanges()}>Endre oppskrift</Button.Success>
      </>
    );
  }

  //legger til nye ingredienser, sjekker først om de finnes, så legger den til i databasen og så blir det hentet ned igjen
  addIngredientFunc(ingred_id: number, recipe_id: number) {
    //sjekker om ingrediensen allerede finnes i oppskriften
    const ifExist = this.recipeContent.map((element) =>
      element.ingred_id == ingred_id ? true : false
    );
    //hvis ingrediensen ikke finnes i oppskriften vil den bli lagt til
    if (!ifExist.includes(true)) {
      const add = { oppskrift_id: recipe_id, ingred_id: ingred_id, mengde: 0, maleenhet: '' };

      //gjennbruker .createRecipeIngredient, den forventer et array så pakker da add inn i et array
      service
        .createRecipeIngredient([add])
        .then(() => this.getIngredRecipe())
        .catch((error) => Alert.danger('Error getting ingredients: ' + error.message));
    } else {
      Alert.info('denne ingrediensen finnes allerede i oppskriften');
    }
  }
  //lager en egen funksjon for å hente ingrediensene til en oppskrift fordi jeg bruker den to steder, praktisk
  getIngredRecipe() {
    service
      .getRecipeContent(this.props.match.params.id)
      .then((recipeContent) => (this.recipeContent = recipeContent))
      .catch((error) => Alert.danger('Error getting recipe content: ' + error.message));
  }
  pushNewChanges() {
    console.log('nå sendes objektet', this.recipeContent);
    service
      .updateRecipe(this.recipe)
      .catch((error) => Alert.danger('Error getting ingredients: ' + error.message));
    service
      .updateRecipeIngredient(this.recipeContent)
      .then(() => history.push('/recipe/' + this.props.match.params.id))
      .catch((error) => Alert.danger('Error getting ingredients: ' + error.message));
  }
  deleteIngredient(recipe_id: number, ingred_id: number) {
    service
      .deleteIngredient(recipe_id, ingred_id)
      .then(() => this.getIngredRecipe())
      .catch((error) => Alert.danger('Error getting ingredients: ' + error.message));
  }
  mounted() {
    service
      .getAllIngredient()
      .then((ingredients) => (this.ingredients = ingredients))
      .catch((error) => Alert.danger('Error getting ingredients: ' + error.message));
    console.log(this.props.match.params.id);

    this.getIngredRecipe();

    service
      .getRecipe(this.props.match.params.id)
      .then((recipe) => {
        this.recipe = recipe[0];
      })
      .catch((error) => Alert.danger('Error getting recipe: ' + error.message));
  }
}
