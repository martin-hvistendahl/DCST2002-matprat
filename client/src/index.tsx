import ReactDOM from 'react-dom';
import * as React from 'react';
import { Component } from 'react-simplified';
import { NavLink, HashRouter, Route } from 'react-router-dom';
import { NavBar, Card, Alert, Column, Row, Form, Button, RecipeView } from './widgets';
import { NewRecipe } from './components';

import service, { Recipe } from './service';
import { createHashHistory } from 'history';

class Menu extends Component {
  render() {
    return (
      <NavBar brand="MatForum">
        <NavBar.Link to="/newrecipe">Ny oppskrift</NavBar.Link>
      </NavBar>
    );
  }
}

class Home extends Component {
  recipes: Recipe[] = [];

  render() {
    return (
      <>
        <Card title="Oppskrifter">
          {this.recipes.map((recipe) => (
            <Row key={recipe.oppskrift_id}>
              <Column>
                <NavLink
                  style={{ textDecoration: 'none', color: 'black' }}
                  to={'/recipe/' + recipe.oppskrift_id}
                >
                  <RecipeView
                    img={recipe.bilde_adr}
                    name={recipe.oppskrift_navn}
                    numbOfPors={recipe.ant_pors}
                  ></RecipeView>
                </NavLink>
              </Column>
            </Row>
          ))}
        </Card>
      </>
    );
  }

  mounted() {
    service
      .getAllRepice()
      .then((recipes) => (this.recipes = recipes))
      .catch((error) => Alert.danger('Error getting tasks: ' + error.message));
  }
}

ReactDOM.render(
  <HashRouter>
    <div>
      <Alert />
      <Menu />
      <Route exact path="/" component={Home} />
      <Route exact path="/newrecipe" component={NewRecipe} />
    </div>
  </HashRouter>,
  document.getElementById('root')
);
