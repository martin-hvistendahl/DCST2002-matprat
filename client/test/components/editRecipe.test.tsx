import * as React from 'react';

import { shallow, mount } from 'enzyme';
import { Alert, Card, Button } from '../../src/widgets';
import { EditRecipe } from '../../src/components/editRecipe';
import service from '../../src/service';

import { createHashHistory } from 'history';

const history = createHashHistory();
jest.mock('../../src/service', () => {
  class Service {
    getAllCountry() {
      return Promise.resolve([
        {
          land_id: 1,
          land_navn: 'Sverige',
        },
        {
          land_id: 2,
          land_navn: 'Italia',
        },
      ]);
    }
    getAllCategory() {
      return Promise.resolve([
        {
          kategori_id: 1,
          kategori_navn: 'Ikea mat',
        },
        {
          kategori_id: 2,
          kategori_navn: 'enkelt',
        },
      ]);
    }
    getAllIngredient() {
      return Promise.resolve([
        {
          ingred_id: 1,
          ingred_navn: 'pizzadeig',
        },
        {
          ingred_id: 2,
          ingred_navn: 'pizza fyll',
        },
        {
          ingred_id: 3,
          ingred_navn: 'kjottboller',
        },
      ]);
    }
    getRecipe(id: number) {
      return Promise.resolve([
        {
          oppskrift_id: 1,
          oppskrift_navn: 'Pizza',
          oppskrift_beskrivelse: 'Pizza er god og enkel',
          oppskrift_steg: 'Bland deigen og la den heve',
          ant_pors: 4,
          bilde_adr: 'pizza.jpg',
          kategori_id: 2,
          land_id: 2,
          ant_like: 1,
        },
      ]);
    }

    getRecipeContent(id: number) {
      return Promise.resolve([
        {
          oppskrift_id: 1,
          ingred_id: 1,
          mengde: 1,
          maleenhet: 'stk',
        },
        {
          oppskrift_id: 1,
          ingred_id: 2,
          mengde: 1,
          maleenhet: 'håndfull',
        },
      ]);
    }
    deleteRecipe(id: number) {
      return Promise.resolve();
    }
    createIngredient() {
      return Promise.resolve();
    }
    createRecipe() {
      return Promise.resolve(1);
    }
    updateRecipe() {
      return Promise.resolve();
    }
    updateRecipeIngredient() {
      return Promise.resolve();
    }
    deleteIngredient() {
      return Promise.resolve();
    }
    createRecipeIngredient(add: []) {
      return Promise.resolve();
    }
  }

  return new Service();
});

describe('editRecipe test', () => {
  //viktig å faktisk sjekke at snapshotet blir riktig
  test('page renders correct with snapshot', (done) => {
    const wrapper = shallow(<EditRecipe match={{ params: { id: 1 } }} />);
    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  test('cancel change', (done) => {
    const wrapper = shallow(<EditRecipe match={{ params: { id: 1 } }} />);
    wrapper.find('#cancelEdit').simulate('click');
    setTimeout(() => {
      expect(window.location.href).toEqual('http://localhost/#/recipe/1');
      done();
    });
  });

  test('input steps on recipe, and portions and confirm change', (done) => {
    const wrapper = shallow(<EditRecipe match={{ params: { id: 1 } }} />);

    wrapper.find('#recipe_step').simulate('change', { currentTarget: { value: 'lag piazz' } });
    wrapper.find('#recipe_portions').simulate('change', { currentTarget: { value: '1' } });

    setTimeout(() => {
      wrapper.find(Button.Success).at(1).simulate('click');
      expect(window.location.href).toEqual('http://localhost/#/recipe/1');
      done();
    });
  });

  test('delete ingredeient from recipe', (done) => {
    const wrapper = shallow(<EditRecipe match={{ params: { id: 1 } }} />);

    setTimeout(() => {
      wrapper.find(Button.Danger).at(0).simulate('click', { oppskrift_id: 1, ingred_id: 1 });

      setTimeout(() => {
        //expect outprintetingredient to be 1
        expect(wrapper.find('#outprintIngredient').length).toBe(1);
        done();
      });
    });
  });

  test('change ingredient inn recipe and save', (done) => {
    //spy on updateRecipeIngredient
    const spy = jest.spyOn(service, 'updateRecipeIngredient');
    const wrapper = shallow(<EditRecipe match={{ params: { id: 1 } }} />);
    setTimeout(() => {
      wrapper.find('#ingredNumber0').simulate('change', { currentTarget: { value: '2' } });
      wrapper.find('#ingredType0').simulate('change', { currentTarget: { value: 'bryst' } });
      setTimeout(() => {
        expect(wrapper.find('#ingredNumber0').prop('value')).toBe('2');
        expect(wrapper.find('#ingredType0').prop('value')).toBe('bryst');
        wrapper.find(Button.Success).at(1).simulate('click');
        //expect updateRecipeIngredient to be called as a spy function
        expect(spy).toHaveBeenCalled();
        done();
      });
    });
  });

  test('add ingredeient to recipe', (done) => {
    const wrapper = mount(<EditRecipe match={{ params: { id: 1 } }} />);
    setTimeout(() => {
      wrapper.find('#choseIngredient').at(0).simulate('keyDown', { key: 'ArrowDown', keyCode: 40 });
      wrapper.find('#choseIngredient').at(0).simulate('keyDown', { key: 'ArrowDown', keyCode: 40 });
      wrapper.find('#choseIngredient').at(0).simulate('keyDown', { key: 'ArrowDown', keyCode: 40 });
      wrapper.find('#choseIngredient').at(0).simulate('keyDown', { key: 'ArrowDown', keyCode: 40 });
      wrapper.find('#choseIngredient').at(0).simulate('keyDown', { key: 'Enter', keyCode: 13 });
      setTimeout(() => {
        expect(wrapper.find('#choseIngredient').at(0).first().text()).toEqual('kjottboller');
        expect(wrapper.containsMatchingElement(<p>kjottboller</p>)).toEqual(true);
        done();
      });
    });
  });
  test('create ingredent sucessfully', (done) => {
    const wrapper = shallow(<EditRecipe match={{ params: { id: 1 } }} />);
    setTimeout(() => {
      wrapper.find('#createIngredient').simulate('change', { currentTarget: { value: 'melk' } });
      wrapper.find('#createIngredientFunc').simulate('click');
      setTimeout(() => {
        expect(wrapper.find('#createIngredient').prop('value')).toBe('');
        //eneste vi får testet her er at vi endrer input feltet, klikker på knappen legg til og så sjekker vi at input feltet er tomt
        //hvis inputfeltet er tomt så er det vel en indikasjon på at ingrediensen ble lagt til fordi funksjonen blir kjørt riktig
        done();
      });
    });
  });
});
