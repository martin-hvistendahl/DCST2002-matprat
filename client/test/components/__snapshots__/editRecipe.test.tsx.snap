// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`editRecipe test page renders correct with snapshot 1`] = `
<Fragment>
  <div
    className="margintop"
  >
    <Row>
      <div
        className="col-6"
        style={
          {
            "paddingRight": "0px",
          }
        }
      >
        <CardFull>
          <h3>
            Pizza
          </h3>
          <br />
          <div
            className="col"
          >
            <Row>
              <div
                className="scroll"
                id="outprintIngredient"
              >
                <Row
                  key="0"
                >
                  <p
                    style={
                      {
                        "width": "215px",
                      }
                    }
                  >
                    pizzadeig
                  </p>
                  <input
                    className="form-control"
                    id="ingredNumber0"
                    onChange={[Function]}
                    style={
                      {
                        "marginRight": "0px",
                        "width": "75px",
                      }
                    }
                    type="number"
                    value={1}
                  />
                  <input
                    className="form-control"
                    id="ingredType0"
                    onChange={[Function]}
                    style={
                      {
                        "width": "120px",
                      }
                    }
                    type="text"
                    value="stk"
                  />
                  <Column
                    width={2}
                  >
                    <ButtonDanger
                      onClick={[Function]}
                    >
                      x
                    </ButtonDanger>
                  </Column>
                </Row>
                <Row
                  key="1"
                >
                  <p
                    style={
                      {
                        "width": "215px",
                      }
                    }
                  >
                    pizza fyll
                  </p>
                  <input
                    className="form-control"
                    id="ingredNumber1"
                    onChange={[Function]}
                    style={
                      {
                        "marginRight": "0px",
                        "width": "75px",
                      }
                    }
                    type="number"
                    value={1}
                  />
                  <input
                    className="form-control"
                    id="ingredType1"
                    onChange={[Function]}
                    style={
                      {
                        "width": "120px",
                      }
                    }
                    type="text"
                    value="håndfull"
                  />
                  <Column
                    width={2}
                  >
                    <ButtonDanger
                      onClick={[Function]}
                    >
                      x
                    </ButtonDanger>
                  </Column>
                </Row>
              </div>
            </Row>
            <br />
            <br />
            <Row>
              <p
                style={
                  {
                    "width": "215px",
                  }
                }
              >
                Ingredienser:
              </p>
              <ForwardRef
                className="selecting"
                id="choseIngredient"
                onChange={[Function]}
                options={
                  [
                    {
                      "label": "Søk på ingredienser",
                      "value": 0,
                    },
                    {
                      "label": "pizzadeig",
                      "value": 1,
                    },
                    {
                      "label": "pizza fyll",
                      "value": 2,
                    },
                    {
                      "label": "kjottboller",
                      "value": 3,
                    },
                  ]
                }
              />
            </Row>
            <br />
            <Row>
              <p
                style={
                  {
                    "width": "215px",
                  }
                }
              >
                Legg til ny:
              </p>
              <FormInput
                id="createIngredient"
                onChange={[Function]}
                placeholder="Skriv inn ny ingrediens"
                style={
                  {
                    "width": "195px",
                  }
                }
                type="text"
                value=""
              />
            </Row>
             
            <Row>
              <p
                style={
                  {
                    "width": "205px",
                  }
                }
              />
              <Column>
                <ButtonSuccess
                  id="createIngredientFunc"
                  onClick={[Function]}
                >
                  Legg til
                </ButtonSuccess>
              </Column>
            </Row>
            <br />
            <Row>
              <p
                style={
                  {
                    "width": "215px",
                  }
                }
              >
                Porsjoner:
              </p>
              <FormInput
                id="recipe_portions"
                onChange={[Function]}
                style={
                  {
                    "width": "195px",
                  }
                }
                type="number"
                value={4}
              />
            </Row>
          </div>
        </CardFull>
      </div>
      <div
        className="col-6"
        style={
          {
            "paddingLeft": "0px",
            "paddingRight": "0px",
          }
        }
      >
        <CardFull>
          <Row>
            <Column>
              <h3>
                Oppskrift:
              </h3>
              <br />
              <Column>
                <FormTextarea
                  id="recipe_step"
                  onChange={[Function]}
                  rows={5}
                  style={
                    {
                      "height": "600px",
                      "width": "600px",
                    }
                  }
                  type="text"
                  value="Bland deigen og la den heve"
                />
              </Column>
            </Column>
          </Row>
        </CardFull>
      </div>
    </Row>
     
    <ButtonSuccess
      onClick={[Function]}
    >
      Endre oppskrift
    </ButtonSuccess>
    <ButtonDanger
      id="cancelEdit"
      onClick={[Function]}
    >
      Avbryt
    </ButtonDanger>
  </div>
</Fragment>
`;
